File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/CSharp/Silverlight/1041/BusinessApplication/Models/RegistrationData.partial.cs
namespace $safeprojectname$.Web
{
    using System;
    using System.ComponentModel.DataAnnotations;
    using System.ServiceModel.DomainServices.Client;
    using System.ServiceModel.DomainServices.Client.ApplicationServices;
    using $safeprojectname$.Web.Resources;

    /// <summary>
    /// クライアント側のカスタム検証および <see cref="RegistrationData"/> へのデータ バインディングを指定するための拡張機能です。
    /// </summary>
    public partial class RegistrationData
    {
        private OperationBase currentOperation;

        /// <summary>
        /// パスワードを返す関数を取得または設定します。
        /// </summary>
        internal Func<string> PasswordAccessor { get; set; }

        /// <summary>
        /// パスワードを取得および設定します。
        /// </summary>
        [Required(ErrorMessageResourceName = "ValidationErrorRequiredField", ErrorMessageResourceType = typeof(ValidationErrorResources))]
        [Display(Order = 3, Name = "PasswordLabel", Description = "PasswordDescription", ResourceType = typeof(RegistrationDataResources))]
        [RegularExpression("^.*[^a-zA-Z0-9].*$", ErrorMessageResourceName = "ValidationErrorBadPasswordStrength", ErrorMessageResourceType = typeof(ValidationErrorResources))]
        [StringLength(50, MinimumLength = 7, ErrorMessageResourceName = "ValidationErrorBadPasswordLength", ErrorMessageResourceType = typeof(ValidationErrorResources))]
        public string Password
        {
            get
            {
                return (this.PasswordAccessor == null) ? string.Empty : this.PasswordAccessor();
            }

            set
            {
                this.ValidateProperty("Password", value);
                this.CheckPasswordConfirmation();

                // パスワードをプレーン テキストでメモリに保存することはできないため、パスワードをプライベート フィールドに保存しないでください。
                // 代わりに、指定された PasswordAccessor が値のバッキング ストアとして機能します。

                this.RaisePropertyChanged("Password");
            }
        }

        /// <summary>
        /// パスワードの確認を返す関数を取得または設定します。
        /// </summary>
        internal Func<string> PasswordConfirmationAccessor { get; set; }

        /// <summary>
        /// パスワードの確認文字列を取得および設定します。
        /// </summary>
        [Required(ErrorMessageResourceName = "ValidationErrorRequiredField", ErrorMessageResourceType = typeof(ValidationErrorResources))]
        [Display(Order = 4, Name = "PasswordConfirmationLabel", ResourceType = typeof(RegistrationDataResources))]
        public string PasswordConfirmation
        {
            get
            {
                return (this.PasswordConfirmationAccessor == null) ? string.Empty : this.PasswordConfirmationAccessor();
            }

            set
            {
                this.ValidateProperty("PasswordConfirmation", value);
                this.CheckPasswordConfirmation();

                // パスワードをプレーン テキストでメモリに保存することはできないため、パスワードをプライベート フィールドに保存しないでください。
                // 代わりに、指定された PasswordAccessor が値のバッキング ストアとして機能します。

                this.RaisePropertyChanged("PasswordConfirmation");
            }
        }

        /// <summary>
        /// 現在の登録操作またはログイン操作を取得または設定します。
        /// </summary>
        internal OperationBase CurrentOperation
        {
            get
            {
                return this.currentOperation;
            }
            set
            {
                if (this.currentOperation != value)
                {
                    if (this.currentOperation != null)
                    {
                        this.currentOperation.Completed -= (s, e) => this.CurrentOperationChanged();
                    }

                    this.currentOperation = value;

                    if (this.currentOperation != null)
                    {
                        this.currentOperation.Completed += (s, e) => this.CurrentOperationChanged();
                    }

                    this.CurrentOperationChanged();
                }
            }
        }

        /// <summary>
        /// ユーザーが現在登録されているか、またはログインしているかを示す値を取得します。
        /// </summary>
        [Display(AutoGenerateField = false)]
        public bool IsRegistering
        {
            get
            {
                return this.CurrentOperation != null && !this.CurrentOperation.IsComplete;
            }
        }

        /// <summary>
        /// 現在の操作が変更された場合のヘルパー メソッドです。
        /// 適切なプロパティの変更通知を送信するために使用されます。
        /// </summary>
        private void CurrentOperationChanged()
        {
            this.RaisePropertyChanged("IsRegistering");
        }

        /// <summary>
        /// パスワードとパスワードの確認が一致することを確認します。
        /// 一致しない場合は、検証エラーが追加されます。
        /// </summary>
        private void CheckPasswordConfirmation()
        {
            // いずれかのパスワードが入力されていない場合、フィールドが同じかどうかのテストは行われません。
            // Required 属性により、両方のフィールドに必ず値が入力されるようになります。
            if (string.IsNullOrWhiteSpace(this.Password)
                || string.IsNullOrWhiteSpace(this.PasswordConfirmation))
            {
                return;
            }

            // 値が異なる場合は、両方のメンバーが指定された検証エラーを追加します
            if (this.Password != this.PasswordConfirmation)
            {
                this.ValidationErrors.Add(new ValidationResult(ValidationErrorResources.ValidationErrorPasswordConfirmationMismatch, new string[] { "PasswordConfirmation", "Password" }));
            }
        }

        /// <summary>
        /// UserName 値の入力後、ロジックを実行します。
        /// </summary>
        /// <param name="userName">入力されたユーザー名の値です。</param>
        /// <remarks>
        /// 値が完全に入力されたことをフォームが示すようにします。
        /// OnUserNameChanged メソッドを使用すると、ユーザーがフォームで値の入力を完了する前に途中で呼び出すことができます。
        /// </remarks>
        internal void UserNameEntered(string userName)
        {
            // フレンドリ名が指定されていない場合は、新しいエンティティの UserName と一致するように FriendlyName が自動入力されます
            if (string.IsNullOrWhiteSpace(this.FriendlyName))
            {
                this.FriendlyName = userName;
            }
        }

        /// <summary>
        /// このエンティティのデータで初期化された新しい <see cref="LoginParameters"/> を作成します (IsPersistent の既定値は false になります)。
        /// </summary>
        public LoginParameters ToLoginParameters()
        {
            return new LoginParameters(this.UserName, this.Password, false, null);
        }
    }
}

File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/VisualBasic/Silverlight/2052/BusinessApplication/Models/RegistrationData.partial.vb
Imports System
Imports System.ComponentModel
Imports System.ComponentModel.DataAnnotations
Imports System.ServiceModel.DomainServices.Client
Imports System.ServiceModel.DomainServices.Client.ApplicationServices

Namespace Web

    Partial Public Class RegistrationData
        Private _CurrentOperation As OperationBase

        ''' <summary>
        ''' 获取或设置返回密码的函数。
        ''' </summary>
        Friend Property PasswordAccessor() As Func(Of String)

        ''' <summary>
        ''' 获取和设置密码。
        ''' </summary>
        <Required(ErrorMessageResourceName:="ValidationErrorRequiredField", ErrorMessageResourceType:=GetType(ValidationErrorResources))> _
        <Display(Order:=3, Name:="PasswordLabel", Description:="PasswordDescription", ResourceType:=GetType(RegistrationDataResources))> _
        <RegularExpression("^.*[^a-zA-Z0-9].*$", ErrorMessageResourceName:="ValidationErrorBadPasswordStrength", ErrorMessageResourceType:=GetType(ValidationErrorResources))> _
        <StringLength(50, MinimumLength:=7, ErrorMessageResourceName:="ValidationErrorBadPasswordLength", ErrorMessageResourceType:=GetType(ValidationErrorResources))> _
        Public Property Password() As String
            Get
                Return If((Me.PasswordAccessor Is Nothing), String.Empty, Me.PasswordAccessor.Invoke())
            End Get

            Set(ByVal value As String)
                Me.ValidateProperty("Password", value)
                Me.CheckPasswordConfirmation()

                ' 请不要将密码存储在私有字段中，因为不应以纯文本形式在内存中存储密码。
                ' 而应将提供的 PasswordAccessor 用作该值的后备存储。

                Me.RaisePropertyChanged("Password")
            End Set
        End Property

        ''' <summary>
        ''' 获取或设置返回密码确认的函数。
        ''' </summary>
        Friend Property PasswordConfirmationAccessor() As Func(Of String)

        ''' <summary>
        ''' 获取和设置密码确认字符串。
        ''' </summary>
        <Required(ErrorMessageResourceName:="ValidationErrorRequiredField", ErrorMessageResourceType:=GetType(ValidationErrorResources))> _
        <Display(Order:=4, Name:="PasswordConfirmationLabel", ResourceType:=GetType(RegistrationDataResources))> _
        Public Property PasswordConfirmation() As String
            Get
                Return If((Me.PasswordConfirmationAccessor Is Nothing), String.Empty, Me.PasswordConfirmationAccessor.Invoke())
            End Get

            Set(ByVal value As String)
                Me.ValidateProperty("PasswordConfirmation", value)
                Me.CheckPasswordConfirmation()

                ' 请不要将密码存储在私有字段中，因为不应以纯文本形式在内存中存储密码。
                ' 而应将提供的 PasswordAccessor 用作该值的后备存储。

                Me.RaisePropertyChanged("PasswordConfirmation")
            End Set
        End Property

        ''' <summary>
        ''' 获取或设置当前注册或登录操作。
        ''' </summary>
        Friend Property CurrentOperation() As OperationBase
            Get
                Return _CurrentOperation
            End Get
            Set(ByVal value As OperationBase)
                If Not Object.Equals(_CurrentOperation, value) Then
                    If Not IsNothing(_CurrentOperation) Then
                        RemoveHandler _CurrentOperation.Completed, AddressOf Me.CurrentOperationChanged
                    End If

                    _CurrentOperation = value

                    If Not IsNothing(_CurrentOperation) Then
                        AddHandler _CurrentOperation.Completed, AddressOf Me.CurrentOperationChanged
                    End If

                    Me.CurrentOperationChanged(Me, EventArgs.Empty)
                End If
            End Set
        End Property

        ''' <summary>
        ''' 获取一个值，该值指示用户当前是否正在注册或登录。
        ''' </summary>
        <Display(AutoGenerateField:=False)> _
        Public ReadOnly Property IsRegistering() As Boolean
            Get
                Return (Not IsNothing(Me.CurrentOperation)) AndAlso (Not Me.CurrentOperation.IsComplete)
            End Get
        End Property

        ''' <summary>
        ''' 在当前操作发生更改时要使用的帮助程序方法。
        ''' 用于引发合适的属性更改通知。
        ''' </summary>
        ''' <param name="sender">事件发送方。</param>
        ''' <param name="e">event 参数。</param>
        Private Sub CurrentOperationChanged(ByVal sender As Object, ByVal e As EventArgs)
            Me.RaisePropertyChanged("IsRegistering")
        End Sub

        ''' <summary>
        ''' 检查以确保密码与确认密码匹配。
        ''' 如果两者不匹配，则添加验证错误。
        ''' </summary>
        Private Sub CheckPasswordConfirmation()
            ' 如果未输入两个密码中的任何一个，则不测试这两个字段的相等性。
            ' Required 特性确保为这两个字段都输入了值。
            If String.IsNullOrWhiteSpace(Me.Password) OrElse String.IsNullOrWhiteSpace(Me.PasswordConfirmation) Then
                Exit Sub
            End If

            ' 如果两个值不同，则添加一个验证错误，并同时指定这两个成员。
            If Me.Password <> Me.PasswordConfirmation Then
                Me.ValidationErrors.Add(New ValidationResult(ValidationErrorResources.ValidationErrorPasswordConfirmationMismatch, New String() {"PasswordConfirmation", "Password"}))
            End If
        End Sub

        ''' <summary>
        ''' 在输入 UserName 值之后执行逻辑
        ''' </summary>
        ''' <param name="userName">输入的用户名。</param>
        ''' <remarks>
        ''' 允许窗体指示值在何时已输入完整。
        ''' 使用 OnUserNameChanged 方法可能导致在用户在窗体中输入完整值之前过早进行调用。
        ''' </remarks>
        Friend Sub UserNameEntered(ByVal userName As String)
            ' 在没有指定友好名称时为新实体自动填充 FriendlyName 以匹配 UserName
            If String.IsNullOrWhiteSpace(Me.FriendlyName) Then
                Me.FriendlyName = userName
            End If
        End Sub

        ''' <summary>
        ''' 创建使用此实体的数据(IsPersistent 默认为 false)初始化的新 <see cref="LoginParameters"/>。
        ''' </summary>
        Public Function ToLoginParameters() As LoginParameters
            Return New LoginParameters(Me.UserName, Me.Password, False, Nothing)
        End Function
    End Class
End Namespace